pkg/experimental/storage/filepath/jsonfile_rest.go

```diff
@@ -0,0 +1,429 @@
+func (f *filepathREST) objectFileName(ctx context.Context, name string) string {
+	if f.isNamespaced {
+		// FIXME: return error if namespace is not found
+		ns, _ := genericapirequest.NamespaceFrom(ctx)
+		return filepath.Join(f.objRootPath, ns, name+".json")
+	}
+	return filepath.Join(f.objRootPath, name+".json")
+}
+
+func (f *filepathREST) objectDirName(ctx context.Context) string {
+	if f.isNamespaced {
+		// FIXME: return error if namespace is not found
+		ns, _ := genericapirequest.NamespaceFrom(ctx)
+		return filepath.Join(f.objRootPath, ns)
+	}
+	return filepath.Join(f.objRootPath)
+}
+
+func write(encoder runtime.Encoder, filepath string, obj runtime.Object) error {
+	buf := new(bytes.Buffer)
+	if err := encoder.Encode(obj, buf); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filepath, buf.Bytes(), 0600)
+}
+
+func read(decoder runtime.Decoder, path string, newFunc func() runtime.Object) (runtime.Object, error) {
+	content, err := ioutil.ReadFile(filepath.Clean(path))
+	if err != nil {
+		return nil, err
+	}
+	newObj := newFunc()
+	decodedObj, _, err := decoder.Decode(content, nil, newObj)
+	if err != nil {
+		return nil, err
+	}
+	return decodedObj, nil
+}
+
+func exists(filepath string) bool {
+	_, err := os.Stat(filepath)
+	return err == nil
+}
+
+func ensureDir(dirname string) error {
+	if !exists(dirname) {
+		return os.MkdirAll(dirname, 0700)
+	}
+	return nil
+}
+
+func visitDir(dirname string, newFunc func() runtime.Object, codec runtime.Decoder, visitFunc func(string, runtime.Object)) error {
+	return filepath.Walk(dirname, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if !strings.HasSuffix(info.Name(), ".json") {
+			return nil
+		}
+		newObj, err := read(codec, path, newFunc)
+		if err != nil {
+			return err
+		}
+		visitFunc(path, newObj)
+		return nil
+	})
+}
+
+func appendItem(v reflect.Value, obj runtime.Object) {
+	v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+}
+
+func getListPrt(listObj runtime.Object) (reflect.Value, error) {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return reflect.Value{}, err
+	}
+	v, err := conversion.EnforcePtr(listPtr)
+	if err != nil || v.Kind() != reflect.Slice {
+		return reflect.Value{}, fmt.Errorf("need ptr to slice: %v", err)
+	}
+	return v, nil
+}
+
+func (f *filepathREST) Watch(ctx context.Context, options *metainternalversion.ListOptions) (watch.Interface, error) {
+	jw := &jsonWatch{
+		id: len(f.watchers),
+		f:  f,
+		ch: make(chan watch.Event, 10),
+	}
+	// On initial watch, send all the existing objects
+	list, err := f.List(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+
+	danger := reflect.ValueOf(list).Elem()
+	items := danger.FieldByName("Items")
+
+	for i := 0; i < items.Len(); i++ {
+		obj := items.Index(i).Addr().Interface().(runtime.Object)
+		jw.ch <- watch.Event{
+			Type:   watch.Added,
+			Object: obj,
+		}
+	}
+
+	// TODO: mutex
```

Comment:
```suggestion
	// TODO: mutex
```
I added the mutex :)

Reply:
remove the TODO